Initial engine framework

fmontei opened the pull request:
This commit adds the initial engine framework for Deckhand. Included
is the logic for parsing YAML files as well as validating them and
doing forward substitution as specified by the YAML file.

This commit also includes unit tests for the framework changes.

TODO: Integrate with Deckhand API to retrieve secret store values for forward substitution.

Commit:
Initial engine framework

This commit adds the initial engine framework for Deckhand. Included
is the logic for parsing YAML files as well as validating them and
doing forward substitution as specified by the YAML file.

This commit also includes unit tests for the framework changes.

mark-burnett commented on deckhand/engine/secret_substitution.py:
```diff
@@ -0,0 +1,126 @@
+        """
+        attrs = multi_key.split('.')
+        # Ignore the first attribute if it is "." as that is a self-reference.
+        if attrs[0] == '':
```
You might consider using https://pypi.python.org/pypi/jsonpath-ng/1.4.2 to help with substitutions.

fmontei replied:
Thanks, I'll look into which is more intuitive and make the updates in a follow-up patch.

mark-burnett reviewed (commented):
We will probably want to use a 3rd party schema validation system, maybe jsonschema:

- [jsonschema docs](http://json-schema.org/)
- [jsonschema python library](https://github.com/Julian/jsonschema)

Commit:
Add jsonschema validation to Deckhand

Previously Deckhand was using manual validation of YAML files.
However, schema validation is much cleaner, robust and thorough.
So, this commit adds jsonschema as a dependence in Deckhand.
Tempest already uses it as a dependence as well -- so a precedence
already exists.

This commit also updates unit tests as needed. Documentation
changes will be made in a follow-up patch.

Commit:
Add additional documentation.

sh8121att commented on deckhand/control/api.py:
```diff
@@ -28,7 +28,7 @@ def start_api(state_manager=None):
 
     v1_0_routes = [
-        ('/secrets', secrets.SecretsResource())
+        ('secrets', secrets.SecretsResource())
```
I'm not sure we'll want an API endpoint specific to secrets, but understand that the API is waiting on better definition of functionality

sh8121att commented on deckhand/engine/schema/v1_0/default_schema.py:
```diff
@@ -0,0 +1,69 @@
+            'properties': {
+                'apiVersion': {
+                    'type': 'string',
+                    'choices': ['deckhand/v1']
```
This should probably be dynamically validated based on loaded schemas.

mark-burnett replied:
Ah, good catch.  We should probably make sure we define the apiVersion validation in one place in the schema and just use it in the multiple places.

fmontei replied:
Fixed in next commit.

sh8121att commented on deckhand/engine/schema/v1_0/default_schema.py:
```diff
@@ -0,0 +1,69 @@
+        'apiVersion': {
+            'type': 'string',
+            # TODO(fm577c): These should be enumerated.
+            'choices': ['service/v1']
```
I don't think we want to enumerate these actually, otherwise deckhand contains intrinsic knowledge of available UCP schemas.

fmontei replied:
Fixed in next commit.

sh8121att commented on deckhand/engine/schema/v1_0/default_schema.py:
```diff
@@ -0,0 +1,69 @@
+        'kind': {
+            'type': 'string',
+            # TODO(fm577c): These should be enumerated.
+            'choices': ['ConsumerOfCertificateData']
```
Again, this probably shouldn't be enumerated in the schema.

fmontei replied:
Fixed in next commit.

mark-burnett commented on deckhand/engine/schema/v1_0/default_schema.py:
```diff
@@ -0,0 +1,69 @@
+    'properties': {
+        'apiVersion': {
+            'type': 'string',
+            # TODO(fm577c): These should be enumerated.
```
Maybe a regex would be adequate validation?  I don't think we'll have Deckhand know every possible schemaVersion/apiVersion.

fmontei replied:
Fixed in next commit.

mark-burnett commented on deckhand/engine/schema/v1_0/default_schema.py:
```diff
@@ -0,0 +1,69 @@
+        },
+        'kind': {
+            'type': 'string',
+            # TODO(fm577c): These should be enumerated.
```
Likewise here, I suspect we will not restrict the list of valid kinds here.  A regex enforcing a title-case format might be ok.

fmontei replied:
Fixed in next commit.

mark-burnett commented on deckhand/engine/schema/v1_0/default_schema.py:
```diff
@@ -0,0 +1,69 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+substitution_schema = {
```
It's more intuitive for me that the schemas would be defined as non-code files.  I think either JSON or YAML would be reasonable.

Then we can specify it completely and include, e.g. the schema header: `"$schema": "http://json-schema.org/schema#"`

fmontei replied:
Well, according to the API docs [0], the jsonchema constructor expects a dictionary as the schema to validate the data against. So the one issue with using a JSON file as the schema is that it would require opening it and reading it, then calling JSON.parse() on it. This is more error-prone than simply importing the schema directly. I personally prefer including the schema in Python code -- which is what Tempest also does (which means there's precedence for it).

[0] https://python-jsonschema.readthedocs.io/en/latest/validate/#jsonschema.IValidator

fmontei replied:
I mean JSON.load()

fmontei replied:
I might have misunderstood your comment. Do you mean adding a JSON ref in the Python schema code to a JSON file? Assuming that's possible.

mark-burnett replied:
I think you understood.  I was just suggesting using a json file to hold the metadata schema.  It's not a critical point, and I agree it's a bit easier to keep it in the python code, so no change is strictly needed.

mark-burnett commented on deckhand/engine/schema/v1_0/default_schema.py:
```diff
@@ -0,0 +1,69 @@
+    'required': ['dest', 'src']
+}
+
+schema = {
```
:+1: I like this approach of validating the whole document, but leaving the data section undefined.

fmontei replied:
I'm going to do manual validation there. I haven't looked into whether jsonschema is capable of doing forward-lookups like regexes can.

mark-burnett replied:
I don't think you will need to do particularly manual validation of the data sections.  The intention is for the schemas of the data section to be specified at runtime via documents either on the filesystem or through api endpoints.

It's not 100% decided whether we will inject these from a centralized place of ownership or whether the individual services will own particular schemas.

There are definitely additional kinds of validation that will not be easily expressed by schemas as you begin implementing layering and substitution (e.g. does the source path exist in the source document), but even then the final validation of a concrete document should be possible through a provided schema.

mark-burnett reviewed (commented):
Good start on validation.  I will hold off on approving until we agree on the metadata format (hopefully today).

Commit:
Use regexes for jsonschema pre-validation

This commit adds the ability to dynamically pick the right
schema to validate the YAML data against, based on the apiVersion
included in the YAML metadata.

This commit also corrects the pre-validation schema by changing
"choices" to "pattern", allowing Deckhand to simply check for a
regex rather than having to enumerate all possible types of data.

fmontei committed:
Trivial fix to default_schema

Added parenthesis that was missing to regex.

mark-burnett commented on deckhand/engine/schema/v1_0/default_schema.py:
```diff
@@ -39,13 +39,11 @@
             'type': 'string',
-            # TODO(fm577c): These should be enumerated.
-            'choices': ['ConsumerOfCertificateData']
+            'pattern': '^([A-Za-z]+$'
```
I think there's a mismatched paren here.

mark-burnett replied:
Apologies for this, you had already fixed it!

sh8121att commented on deckhand/engine/secret_substitution.py:
```diff
@@ -0,0 +1,157 @@
+        self._validate_with_schema()
+
+        # Validate that each "dest" field exists in the YAML data.
+        substitutions = self.data['metadata']['substitutions']
```
We should just insert data into the document based on the dest field (even if that is means inserting a narrow, deep object) and then allow the final schema validation determine if that dest field is valid.

sh8121att reviewed (commented):
We can merge this in knowing that as we finalize the document format, changes will be made in the future.

Commit:
Update default_schema with our updated schema definition

This commit updates the default_schema based on recent changes
to the document definition that Deckhand is supposed to validate.

Commit:
Remove apiVersion attribute from substitutions.src attributes

The src apiVersion attribute will be dynamically validated based
on the loaded schemas.

Commit:
Remove apiVersion attribute from substitutions.src attributes

The src apiVersion attribute will be dynamically validated based
on the loaded schemas.

Commit:
Merge branch 'init-yaml-framework' of https://github.com/att-comdev/deckhand into init-yaml-framework

mark-burnett commented on deckhand/engine/schema/v1_0/default_schema.py:
```diff
@@ -28,16 +28,12 @@
         'src': {
             'type': 'object',
             'properties': {
-                'apiVersion': {
```
I think we can keep this basic form validation, but you're right that we should validate against known schema names as well (later once that's in place).

mark-burnett commented:
@fmontei If you can fix the conflict in requirements.txt, I'm happy to merge this.

fmontei committed:
Merge branch 'master' into init-yaml-framework

fmontei commented:
@mark-burnett 

Thanks, resolved conflict in requirements.txt.

Commit:
Change apiVersion references to schemaVersion

Commit:
Merge branch 'init-yaml-framework' of https://github.com/att-comdev/deckhand into init-yaml-framework

Commit:
Add kind param to SchemaVersion class.
